overlord/restart/restart.go

```diff
@@ -362,3 +362,70 @@ func ReplaceBootID(st *state.State, bootID string) {
 	rm := restartManager(st, "internal error: cannot mock a restart request before RestartManager initialization")
 	rm.bootID = bootID
 }
+
+// TaskWaitForRestart can be used for tasks that need to wait for a pending
+// restart to occur. The task will then be re-run after the restart has occurred.
+func TaskWaitForRestart(t *state.Task) error {
+	chg := t.Change()
+	rt, err := changeRestartContext(chg)
+	if err != nil {
+		return err
+	}
+	if rt == nil {
+		return nil
+	}
+
+	// We catch them in Undoing/Doing and restore them to
+	// Do/Undo so they are re-run as we cannot save progress mid-task.
+	status := t.Status()
+	switch status {
+	case state.UndoingStatus:
+		err = rt.doUndoReboot(t, state.UndoStatus)
+	case state.DoingStatus:
+		err = rt.doReboot(t, state.DoStatus)
+	default:
```

Comment:
which use cases we have for this? I thought the only case we had for setting something in Doing to Wait was the check-rerefresh, this seems extremely complex if that is the only case?